File: disposia-engine/src/main/java/io/disposia/engine/mapper/UrlMapper.java
package io.disposia.engine.mapper;

import java.net.MalformedURLException;
import java.net.URL;

/**
 * @author Maximilian Irro
 */
public class UrlMapper {

    public static UrlMapper INSTANCE = new UrlMapper();

    public String asString(URL url) {
        return url != null ? url.toExternalForm() : null;
    }

    public URL asURL(String url) {
        try {
            return url != null ? new URL(url) : null;
        } catch (MalformedURLException e) {
            throw new RuntimeException( e );
        }
    }
}

File: disposia-engine/src/main/scala/io/disposia/engine/util/mapper/UrlMapper.scala
package io.disposia.engine.util.mapper

import java.net.{MalformedURLException, URL, URLEncoder}

object UrlMapper {

  def asString(url: URL): Option[String] = Option(url.toExternalForm)

  def asUrl(url: String): Option[URL] =
    try
      Option(url).map(u => new URL(u))
    catch {
      case e: MalformedURLException => throw new RuntimeException(e)
    }

  def sanitize(url: String): Option[String] = Option(url)
    .map(URLEncoder.encode(_, "UTF-8"))
    .map { _
      .replace("%", "")   // the char `^` breaks Play, even URL encoded
      .replace("<", "")
      .replace(">", "")
      .replace("\n", "")
      .replace("\t", "")
      .replace("\r", "")
    }

}
